### PR TITLE
Use apply_channel in DensityMatrixSimulator

### PR DESCRIPTION
**Pull request by Strilanc**
And make it work on operations, not just gate operations.

Fixes https://github.com/quantumlib/Cirq/issues/1630

### CODE REVIEW

File: cirq/sim/density_matrix_simulator.py [hunk 1]
```diff
@@ -30,6 +30,14 @@
     from typing import Any, Hashable
 
 
+class _StateAndBuffers:
+
+    def __init__(self, n: int, state: np.ndarray):
```
Comment (dabacon) [inline cirq/sim/density_matrix_simulator.py]: num_qubits?  or add doc about what n is.

Comment (dabacon) [inline cirq/sim/density_matrix_simulator.py]: i'd call the inner state "matrix" to remind people it's the density matrix, but that's a nit.

Comment (Strilanc) [reply cirq/sim/density_matrix_simulator.py]: Done.

Comment (Strilanc) [reply cirq/sim/density_matrix_simulator.py]: Done.

**Pull request merged**